test/es-module/test-esm-resolve-type.js

```diff
@@ -98,86 +100,148 @@ try {
+      'module.exports = {esm-resolve-tester: () => {return 42}}'
+    );
+
+    // test the resolve
+    const resolveResult = resolve(`${moduleName}`);
+    assert.strictEqual(resolveResult.format, 'module');
+    assert.ok(resolveResult.url.includes('my-dual-package/es/index.js'));
+  }
+
+  testDualPackageWithJsMainScriptAndModuleType();
+
+  // TestParameters are ModuleName, mainRequireScript, mainImportScript,
+  // mainPackageType, subdirPkgJsonType, expectedResolvedFormat, mainSuffix
+  [
+    [ 'mjs-mod-mod', 'index.js', 'index.mjs', 'module', 'module', 'module'],
+    [ 'mjs-com-com', 'idx.js', 'idx.mjs', 'commonjs', 'commonjs', 'module'],
+    [ 'mjs-mod-com', 'index.js', 'imp.mjs', 'module', 'commonjs', 'module'],
+    [ 'js-com-com', 'index.js', 'imp.js', 'commonjs', 'commonjs', 'commonjs'],
+    [ 'js-com-mod', 'index.js', 'imp.js', 'commonjs', 'module', 'module'],
+    [ 'qmod', 'index.js', 'imp.js', 'commonjs', 'module', 'module', '?k=v'],
+    [ 'hmod', 'index.js', 'imp.js', 'commonjs', 'module', 'module', '#Key'],
+    [ 'qhmod', 'index.js', 'imp.js', 'commonjs', 'module', 'module', '?k=v#h'],
+    [ 'ts-mod-com', 'index.js', 'imp.ts', 'module', 'commonjs', undefined],
+  ].forEach((testVariant) => {
+    const [
+      moduleName,
+      mainRequireScript,
+      mainImportScript,
+      mainPackageType,
+      subdirPackageType,
+      expectedResolvedFormat,
+      mainSuffix = '' ] = testVariant;
+
+    const mDir = rel(`node_modules/${moduleName}`);
+    const subDir = rel(`node_modules/${moduleName}/subdir`);
+    const pkg = rel(`node_modules/${moduleName}/package.json`);
+    const subdirPkg = rel(`node_modules/${moduleName}/subdir/package.json`);
+    const esScript = rel(`node_modules/${moduleName}/subdir/${mainImportScript}`);
+    const cjsScript = rel(`node_modules/${moduleName}/subdir/${mainRequireScript}`);
+
+    createDir(nmDir);
+    createDir(mDir);
+    createDir(subDir);
+
+    const mainPkgJsonContent = {
+      type: mainPackageType,
+      main: `./subdir/${mainRequireScript}`,
+      exports: {
+        '.': {
+          'require': `./subdir/${mainRequireScript}${mainSuffix}`,
+          'import': `./subdir/${mainImportScript}${mainSuffix}`
+        },
+        './package.json': './package.json',
+      }
+    };
+    const subdirPkgJsonContent = {
+      type: `${subdirPackageType}`
+    };
+
+    fs.writeFileSync(pkg, JSON.stringify(mainPkgJsonContent));
+    fs.writeFileSync(subdirPkg, JSON.stringify(subdirPkgJsonContent));
+    fs.writeFileSync(
+      esScript,
+      'export function esm-resolve-tester() {return 42}'
+    );
+    fs.writeFileSync(
+      cjsScript,
+      'module.exports = {esm-resolve-tester: () => {return 42}}'
+    );
+
+    // test the resolve
+    const resolveResult = resolve(`${moduleName}`);
+    assert.strictEqual(resolveResult.format, expectedResolvedFormat);
+    assert.ok(resolveResult.url.endsWith(`${moduleName}/subdir/${mainImportScript}${mainSuffix}`));
```

Comment:
I saw that this was previously using `.endsWith()` and Antoine suggested `.includes()` instead (presumably because of the query params in some of the specifiers). I think it might be better to create a `URL` instance and then check `pathname.endsWith()`.

As-is, this test would return a bad result if the query param was that path but the actual pathname was something else:

```js
'foo.js?original=my-dual-package/es/index.js'
  .includes('my-dual-package/es/index.js') // erroneously true
```

I think one of the existing cases' query params should be updated to cover this (instead of `?k=v`).

```suggestion

    const url = new URL(resolveResult.url);
    assert.ok(url.pathname.endsWith(`${moduleName}/subdir/${mainImportScript}${mainSuffix}`));
```

Reply:
This could maybe be a follow-up 🤔

Reply:
I think it was a the other way around. I had it with `.includes()` and @aduh95 proposed the amendment to `.endsWith()` 
 (_which is stricter and better I think_) and at the end it should include the query and/or the hash (`endsWith([...]${mainSuffix})`)in the resolved string. I think in the current form, the test covers the use case the best way possible.

Reply:
Ahhh, yes, sorry, I see now (misunderstood what it was doing).